express/properties/non_scalar/file_content.py

```diff
@@ -8,25 +8,29 @@ class FileContent(NonScalarProperty):
     """
     file_content property.
     Args:
-        **kwargs: required fields are:
-                    - basename (str): the file name, without the path
-                    - filetype (str): text or image
-                    - upload_dir (str): Path Express sees to the working directory where the job file exists
-                    - object_storage_data (dict): describes the object in the bucket, with keys:
-                        - CONTAINER (str): the name of the object store bucket
-                        - NAME (str): the object name in the container
-                        - PROVIDER (str): the cloud provider where the container exists
-                        - REGION (str): the cloud provider region in which the container exists
+        name (str): Name of the property
+        parser (Any): Express parser being used
+        basename (str): the file name, without the path
+        filetype (str): text or image
+        upload_dir (str): Path Express sees to the working directory where the job file exists
+        object_storage_data (dict): describes the object in the bucket, with keys:
+            - CONTAINER (str): the name of the object store bucket
+            - NAME (str): the object name in the container
+            - PROVIDER (str): the cloud provider where the container exists
+            - REGION (str): the cloud provider region in which the container exists
     """
 
-    def __init__(self, name, parser, *args, **kwargs):
+    def __init__(self, name: str, parser: Any,
+                 *args,
+                 basename: str, filetype: str, upload_dir: str, object_storage_data: Dict[str, str],
```

Comment:
Doing this removes `basename`, `filetype`, `upload_dir`, and `object_storage_data` from `**kwargs`, which means they don't get passed to the constructor on line 27, and don't end up in the `self.kwargs` property in the parent constructor.  
```
>>> def whatever(a, b, c, *args, d, e, f, **kwargs):
...   print(args)
...   print(kwargs)
...   print(a, b, c, d, e, f)
...
>>> whatever(1,2,3,4,5,6,7,8,9, d='d', e='e', f='f', g='g', h='h', i='i')
(4, 5, 6, 7, 8, 9)
{'g': 'g', 'h': 'h', 'i': 'i'}
1 2 3 d e f
```
Looking at the code, I can't see an obvious reason for storing the constructor's arguments as properties, it looks like they're only accessed in constructors anyway.  But this is a change in behavior which might trip us up in the future.

I wonder if we'd be better off hedging and doing something like this here, on line 27:
```
passed_kwargs = {'basename': basename, 'filetype': filetype, 'upload_dir': upload_dir, 'object_storage_data': object_storage_data, **kwargs}
super().__init(name, parser, *args, **passed_kwargs)
```

Or maybe it doesn't matter, and there's no good reason to keep the `self.args` and `self.kwargs` properties comprehensive, if we're only accessing these properties here in this class by name anyway.

Reply:
Personally, I think we overuse `*args` and `**kwargs`.  That pattern is basically saying "I abdicate responsibility for specifying this interface".  

Using that feature should really be kept to very specific cases where the inputs aren't distinguished from each other and you can treat them all interchangeably, like `def sum_numbers(*args)` or `def set_environment_variables(**kwargs)`.

Reply:
Agreed on reducing the use of `*args` and `**kwargs` to be more explicit. It seems like the main failure mode this PR might create is one where we're expecting an argument to the init to be passed as `args` or `kwargs`, except it doesn't get passed.

Currently, we hit that failure mode anyway if we forget to give `args` or `kwargs` something - it just gets deferred until we bubble up to whatever nth-grandparent in the inheritance chain tries using it, and we don't actually realize it until we fail a test. If anything, doing it this way at least helps the IDE catch a messed-up function signature.